blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitCacheCloneReadSaveRequest.java

```diff
@@ -0,0 +1,247 @@
+    private static final String REMOTE_REF_BASE = "refs/heads/";
+
+    public GitCacheCloneReadSaveRequest(GitSCMSource gitSource, String branch, String commitMessage, String sourceBranch, String filePath, byte[] contents) {
+        super(gitSource, branch, commitMessage, sourceBranch, filePath, contents);
+    }
+
+    @Override
+    byte[] read() throws IOException {
+        try {
+            GitSCMFileSystem fs = getFilesystem();
+            return fs.invoke(new GitSCMFileSystem.FSFunction<byte[]>() {
+                @Override
+                public byte[] invoke(Repository repository) throws IOException, InterruptedException {
+                    Git activeRepo = getActiveRepository(repository);
+                    Repository repo = activeRepo.getRepository();
+                    File repoDir = repo.getDirectory().getParentFile();
+                    try {
+                        File f = new File(repoDir, filePath);
+                        if (f.canRead()) {
+                            return IOUtils.toByteArray(new FileInputStream(f));
+                        }
+                        return null;
+                    } finally {
+                        FileUtils.deleteDirectory(repoDir);
+                    }
+                }
+            });
+        } catch (InterruptedException ex) {
+            throw new ServiceException.UnexpectedErrorException("Unable to read " + filePath, ex);
+        }
+    }
+
+    @Override
+    void save() throws IOException {
+        try {
+            GitSCMFileSystem fs = getFilesystem();
+            fs.invoke(new GitSCMFileSystem.FSFunction<Void>() {
+                @Override
+                public Void invoke(Repository repository) throws IOException, InterruptedException {
+                    Git activeRepo = getActiveRepository(repository);
+                    Repository repo = activeRepo.getRepository();
+                    File repoDir = repo.getDirectory().getParentFile();
+                    log.fine("Repo cloned to: " + repoDir.getCanonicalPath());
+                    try {
+                        File f = new File(repoDir, filePath);
+                        if (!f.exists() || f.canWrite()) {
+                            try (Writer w = new OutputStreamWriter(new FileOutputStream(f), "utf-8")) {
+                                w.write(new String(contents, "utf-8"));
+                            }
+
+                            try {
+                                AddCommand add = activeRepo.add();
+                                add.addFilepattern(filePath);
+                                add.call();
+
+                                CommitCommand commit = activeRepo.commit();
+                                commit.setMessage(commitMessage);
+                                commit.call();
+
+                                StandardCredentials credential = null;
+                                if (GitUtils.isSshUrl(gitSource.getRemote())) {
```

Comment:
I see similar block of code other places, better to pull this in a util. Also, what expected if its not ssh url? Looks like follow up code will use null credential in that case which probably is ok security wise. For example if git server is setup with username/password, git operations might fail.